core/rsyslog/config.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package rsyslog
+
+import (
+	"crypto/tls"
+	"net/url"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/cert"
+)
+
+var ErrNotConfigured = errors.Errorf("rsyslog fowarding not configured")
+
+func IsNotConfigured(err error) bool {
+	return errors.Cause(err) == ErrNotConfigured
+}
+
+// ClientConfig holds information needed to connect to a
+// rsyslog server.
+type ClientConfig struct {
+	URL    string
+	CACert string
+	Cert   string
+	Key    string
+}
+
+// Validate validates the config values.
+func (c *ClientConfig) Validate() error {
+	if c.URL == "" || c.CACert == "" || c.Cert == "" || c.Key == "" {
+		return errors.Trace(ErrNotConfigured)
+	}
+	u, err := url.Parse(c.URL)
+	if err != nil {
+		return errors.NewNotValid(err, "cannot parse rsyslog url")
+	}
+	if u.Scheme != "https" {
+		return errors.NotValidf("url schema is not https: %v", u.Scheme)
+	}
+	_, err = cert.ParseCert(c.CACert)
+	if err != nil {
+		return errors.NewNotValid(err, "cannot parse rsyslog CA cert")
+	}
+	_, err = cert.ParseCert(c.Cert)
+	if err != nil {
+		return errors.NewNotValid(err, "cannot parse rsyslog client cert")
+	}
+	_, err = tls.X509KeyPair([]byte(c.Cert), []byte(c.Key))
+	if err != nil {
+		return errors.NewNotValid(err, "invalid rsyslog client cert/key pair")
+	}
+	return nil
```

Comment:
it's quite nice for a NotValid to use exact field names, and follow the "XXX not valid" convention where possible. The NotValid bits maybe aren't quite so nice here, but how about:

    u, err := url.Parse(c.URL)
    if err == nil && u.Scheme != "https" {
        err = errors.New("https required")
    }
    if err != nil {
        message := fmt.Sprintf("URL %q not valid", c.URL)
        return errors.NewNotValid(err, message)
    }
    if _, err := cert.ParseCert(c.CACert); err != nil {
        return errors.NewNotValid(err, "CACert not valid")
    }
    if _, err := cert.ParseCert(c.Cert); err != nil {
        return errors.NewNotValid(err, "Cert not valid")
    }
    if _, err := tls.X509KeyPair([]byte(c.Cert), []byte(c.Key)); err != nil {
        return errors.NewNotValid(err, Cert/Key pair not valid)
    }